### PR TITLE
Estimate message fee

### PR DESCRIPTION
**Pull request by joshklop**
Resolves: #681

### CODE REVIEW

File: rpc/handlers.go [hunk 1]
```diff
@@ -1126,3 +1128,22 @@ func (h *Handler) EstimateFee(broadcastedTxns []BroadcastedTransaction, id Block
...
+			Type:               TxnL1Handler,
+			ContractAddress:    msg.To,
+			EntryPointSelector: msg.Selector,
+			CallData:           &msg.Payload,
+			Version: new(felt.Felt), // Needed for transaction hash calculation.
+			Nonce: new(felt.Felt), // Needed for transaction hash calculation.
+		},
+		PaidFeeOnL1: new(felt.Felt), // Needed to marshal to blockifier type.
+	}
+	estimates, rpcErr := h.EstimateFee([]BroadcastedTransaction{tx}, id)
```
Comment (omerfirmak) [inline rpc/handlers.go:1144]: Wouldn't using `vm.Execute` directly instead be easier?

Comment (joshklop) [reply rpc/handlers.go:1144]: Probably. Just hacked this together to get it working. Will refactor once I can reliably execute L1 handler transactions.

**Comment by joshklop** [comment]
State of this PR:

- This correctly executes all L1 handler transactions on goerli2 up to and including block 125,214, with a few exceptions.
- Out of the thousands of L1 handlers, there are eight exceptions that are all Cairo VM-related errors.
- Unfortunately, blockifier only returns the default fee (see https://github.com/starkware-libs/blockifier/issues/734) for successful executions. We can force it to return an error message that contains the correct actual fee. That would mean we call `vm.Execute` twice: once to force an error to get the actual fee, then correctly to get the rest of the `FeeEstimate` fields. That seems kind of excessive.

File: rpc/handlers.go (hunk 1 above)
Comment (joshklop) [reply rpc/handlers.go:1149]: On second thought, I think it makes sense to call `h.EstimateFee`. Almost all the other logic would be repeated otherwise.

**Comment by joshklop** [comment]
This is pretty much done I think, just requires testing. I'll plan to write a script that stress-tests this against real L1 handlers tomorrow.

**Comment by joshklop** [comment]
Tested this against all L1 handlers in the Goerli 2 snapshot. All pass except eight (8), pasted below. The code for the integration test can be found on the `joshklop/estimate-msg-fee-test` branch.

```
18:24:03.349 19/07/2023 -07:00  ERROR   junotest/estimate_message_fee_test.go:157       Failed to estimate message fee  {"errCode": 40, "errMsg": "Contract error", "errData": "failed txn T
rans
actionHash(StarkFelt(\"0x07d2966d833aa5cef7931dd52c90e1a2f81bcd1f222b5a2e1381374fb9f17775\")) reason:ExecutionError(VirtualMachineExecutionErrorWithTrace { trace: \"Error in the called con
trac
t (0x063d1baa3f186d2736af28a471a3aa9bdde3c014f5ae1babfc5350c99860cc93):\\nError at pc=0:27:\\nGot an exception while executing a hint: Custom Hint Error: Class with hash ClassHash(\\n    S
tark
Felt(\\n        \\\"0x04cbbc9d92f7cc0aec678fb706818ba1b43d9deb39409db8e5c1f62e46c0f76a\\\",\\n    ),\\n) is not declared.\\nCairo traceback (most recent call last):\\nUnknown location (pc=
0:23
9)\\nUnknown location (pc=0:227)\\n\", source: CairoRunError(VmException(VmException { pc: 27, inst_location: None, inner_exc: Hint((0, CustomHint(\"Class with hash ClassHash(\\n    StarkF
elt(
\\n        \\\"0x04cbbc9d92f7cc0aec678fb706818ba1b43d9deb39409db8e5c1f62e46c0f76a\\\",\\n    ),\\n) is not declared.\"))), error_attr_value: None, traceback: Some(\"Cairo traceback (most r
ecen
t call last):\\nUnknown location (pc=0:239)\\nUnknown location (pc=0:227)\\n\") })) })"}
github.com/NethermindEth/juno/junotest.getEstimate
        /home/user/repos/juno/junotest/estimate_message_fee_test.go:157
github.com/NethermindEth/juno/junotest.TestEstimateMessageFee.func2
        /home/user/repos/juno/junotest/estimate_message_fee_test.go:99
github.com/sourcegraph/conc/stream.(*Stream).Go.func1
        /home/user/.go/pkg/mod/github.com/sourcegraph/conc@v0.2.0/stream/stream.go:84
github.com/sourcegraph/conc/pool.(*Pool).worker
        /home/user/.go/pkg/mod/github.com/sourcegraph/conc@v0.2.0/pool/pool.go:137
github.com/sourcegraph/conc/panics.(*Catcher).Try
        /home/user/.go/pkg/mod/github.com/sourcegraph/conc@v0.2.0/panics/panics.go:23
github.com/sourcegraph/conc.(*WaitGroup).Go.func1
        /home/user/.go/pkg/mod/github.com/sourcegraph/conc@v0.2.0/waitgroup.go:32
18:24:05.542 19/07/2023 -07:00  ERROR   junotest/estimate_message_fee_test.go:157       Failed to estimate message fee  {"errCode": 40, "errMsg": "Contract error", "errData": "failed txn T
rans
actionHash(StarkFelt(\"0x01e41af6ac269b2d80ad3df0f0540b507c25c2f7a733f2859eb53e71b6797539\")) reason:ExecutionError(VirtualMachineExecutionErrorWithTrace { trace: \"Error in the called con
trac
t (0x0407d7441475afa3d50c70a511ab97f8face4d5b8e616b51455c34ece92a2d16):\\nError at pc=0:27:\\nGot an exception while executing a hint.\\nCairo traceback (most recent call last):\\nUnknown 
loca
tion (pc=0:239)\\nUnknown location (pc=0:227)\\n\\nError in the called contract (0x0407d7441475afa3d50c70a511ab97f8face4d5b8e616b51455c34ece92a2d16):\\nError at pc=0:149:\\nGot an exceptio
n wh
ile executing a hint.\\nCairo traceback (most recent call last):\\nUnknown location (pc=0:2314)\\nUnknown location (pc=0:2233)\\nUnknown location (pc=0:791)\\n\\nError in the called contra
ct (
0x03b960d41dfbe13c9f0c712d81627cf58dc3538180b4e95f9cbec50a29985e80):\\nError at pc=0:15:\\nGot an exception while executing a hint: Custom Hint Error: Class with hash ClassHash(\\n    Star
kFel
t(\\n        \\\"0x00a8ac2bfe98a438946aaca054c0e033923746349b42a5648c85bdd514f30022\\\",\\n    ),\\n) is not declared.\\nCairo traceback (most recent call last):\\nUnknown location (pc=0:2
14)\
\nUnknown location (pc=0:200)\\n\", source: CairoRunError(VmException(VmException { pc: 27, inst_location: None, inner_exc: Hint((0, CustomHint(\"Error in the called contract (0x0407d74414
75af
a3d50c70a511ab97f8face4d5b8e616b51455c34ece92a2d16):\\nError at pc=0:149:\\nGot an exception while executing a hint.\\nCairo traceback (most recent call last):\\nUnknown location (pc=0:231
4)\\
nUnknown location (pc=0:2233)\\nUnknown location (pc=0:791)\\n\\nError in the called contract (0x03b960d41dfbe13c9f0c712d81627cf58dc3538180b4e95f9cbec50a29985e80):\\nError at pc=0:15:\\nGo
t an
 exception while executing a hint: Custom Hint Error: Class with hash ClassHash(\\n    StarkFelt(\\n        \\\"0x00a8ac2bfe98a438946aaca054c0e033923746349b42a5648c85bdd514f30022\\\",\\n  
  ),
\\n) is not declared.\\nCairo traceback (most recent call last):\\nUnknown location (pc=0:214)\\nUnknown location (pc=0:200)\\n\"))), error_attr_value: None, traceback: Some(\"Cairo traceb
ack 
(most recent call last):\\nUnknown location (pc=0:239)\\nUnknown location (pc=0:227)\\n\") })) })"}
github.com/NethermindEth/juno/junotest.getEstimate
        /home/user/repos/juno/junotest/estimate_message_fee_test.go:157
github.com/NethermindEth/juno/junotest.TestEstimateMessageFee.func2
        /home/user/repos/juno/junotest/estimate_message_fee_test.go:99
github.com/sourcegraph/conc/stream.(*Stream).Go.func1
        /home/user/.go/pkg/mod/github.com/sourcegraph/conc@v0.2.0/stream/stream.go:84
github.com/sourcegraph/conc/pool.(*Pool).worker
        /home/user/.go/pkg/mod/github.com/sourcegraph/conc@v0.2.0/pool/pool.go:137
github.com/sourcegraph/conc/panics.(*Catcher).Try
        /home/user/.go/pkg/mod/github.com/sourcegraph/conc@v0.2.0/panics/panics.go:23
github.com/sourcegraph/conc.(*WaitGroup).Go.func1
        /home/user/.go/pkg/mod/github.com/sourcegraph/conc@v0.2.0/waitgroup.go:32
18:26:28.231 19/07/2023 -07:00  ERROR   junotest/estimate_message_fee_test.go:157       Failed to estimate message fee  {"errCode": 40, "errMsg": "Contract error", "errData": "failed txn T
rans
actionHash(StarkFelt(\"0x00489a34d273f759acc646bff44df9f3439d2df5381bb8e3388c2bc059884717\")) reason:ExecutionError(VirtualMachineExecutionErrorWithTrace { trace: \"Error in the called con
trac
t (0x0011559d81452bbdecc84828746c9d7832a7968d0fb09c5f0813e68beab4861f):\\nError at pc=0:24:\\nGot an exception while executing a hint: Assertion failed, 0 % 0x80000000000001100000000000000
0000
000000000000000000000000000001 is equal to 0\\nCairo traceback (most recent call last):\\nUnknown location (pc=0:3497)\\nUnknown location (pc=0:3480)\\nError message: Starkway: Token is no
t in
itialized\\nUnknown location (pc=0:3562)\\n\", source: CairoRunError(VmException(VmException { pc: 24, inst_location: None, inner_exc: Hint((0, AssertNotZero((0, \"0x8000000000000110000000
0000
0000000000000000000000000000000000001\")))), error_attr_value: None, traceback: Some(\"Cairo traceback (most recent call last):\\nUnknown location (pc=0:3497)\\nUnknown location (pc=0:3480
)\\n
Error message: Starkway: Token is not initialized\\nUnknown location (pc=0:3562)\\n\") })) })"}
github.com/NethermindEth/juno/junotest.getEstimate
        /home/user/repos/juno/junotest/estimate_message_fee_test.go:157
github.com/NethermindEth/juno/junotest.TestEstimateMessageFee.func2
        /home/user/repos/juno/junotest/estimate_message_fee_test.go:99
github.com/sourcegraph/conc/stream.(*Stream).Go.func1
        /home/user/.go/pkg/mod/github.com/sourcegraph/conc@v0.2.0/stream/stream.go:84
github.com/sourcegraph/conc/pool.(*Pool).worker
        /home/user/.go/pkg/mod/github.com/sourcegraph/conc@v0.2.0/pool/pool.go:137
github.com/sourcegraph/conc/panics.(*Catcher).Try
        /home/user/.go/pkg/mod/github.com/sourcegraph/conc@v0.2.0/panics/panics.go:23
github.com/sourcegraph/conc.(*WaitGroup).Go.func1
        /home/user/.go/pkg/mod/github.com/sourcegraph/conc@v0.2.0/waitgroup.go:32
18:26:28.271 19/07/2023 -07:00  ERROR   junotest/estimate_message_fee_test.go:157       Failed to estimate message fee  {"errCode": 40, "errMsg": "Contract error", "errData": "failed txn T
rans
actionHash(StarkFelt(\"0x008567e5bcd2b368c4d90ee6e8c222d4dc20a663898e21743658995d6529eee9\")) reason:ExecutionError(VirtualMachineExecutionErrorWithTrace { trace: \"Error in the called con
trac
t (0x0011559d81452bbdecc84828746c9d7832a7968d0fb09c5f0813e68beab4861f):\\nError at pc=0:24:\\nGot an exception while executing a hint: Assertion failed, 0 % 0x80000000000001100000000000000
0000
000000000000000000000000000001 is equal to 0\\nCairo traceback (most recent call last):\\nUnknown location (pc=0:3497)\\nUnknown location (pc=0:3480)\\nError message: Starkway: Token is no
t in
itialized\\nUnknown location (pc=0:3562)\\n\", source: CairoRunError(VmException(VmException { pc: 24, inst_location: None, inner_exc: Hint((0, AssertNotZero((0, \"0x8000000000000110000000
0000
0000000000000000000000000000000000001\")))), error_attr_value: None, traceback: Some(\"Cairo traceback (most recent call last):\\nUnknown location (pc=0:3497)\\nUnknown location (pc=0:3480
)\\n
Error message: Starkway: Token is not initialized\\nUnknown location (pc=0:3562)\\n\") })) })"}
github.com/NethermindEth/juno/junotest.getEstimate
        /home/user/repos/juno/junotest/estimate_message_fee_test.go:157
github.com/NethermindEth/juno/junotest.TestEstimateMessageFee.func2
        /home/user/repos/juno/junotest/estimate_message_fee_test.go:99
github.com/sourcegraph/conc/stream.(*Stream).Go.func1
        /home/user/.go/pkg/mod/github.com/sourcegraph/conc@v0.2.0/stream/stream.go:84
github.com/sourcegraph/conc/pool.(*Pool).worker
        /home/user/.go/pkg/mod/github.com/sourcegraph/conc@v0.2.0/pool/pool.go:137
github.com/sourcegraph/conc/panics.(*Catcher).Try
        /home/user/.go/pkg/mod/github.com/sourcegraph/conc@v0.2.0/panics/panics.go:23
github.com/sourcegraph/conc.(*WaitGroup).Go.func1
        /home/user/.go/pkg/mod/github.com/sourcegraph/conc@v0.2.0/waitgroup.go:32
18:26:31.023 19/07/2023 -07:00  ERROR   junotest/estimate_message_fee_test.go:157       Failed to estimate message fee  {"errCode": 40, "errMsg": "Contract error", "errData": "failed txn T
rans
actionHash(StarkFelt(\"0x0790d398f83c9e5b52be5af23905591768775799abccdb1b0b6e53d36dff3403\")) reason:ExecutionError(VirtualMachineExecutionErrorWithTrace { trace: \"Error in the called con
trac
t (0x0011559d81452bbdecc84828746c9d7832a7968d0fb09c5f0813e68beab4861f):\\nError at pc=0:24:\\nGot an exception while executing a hint: Assertion failed, 0 % 0x80000000000001100000000000000
0000
000000000000000000000000000001 is equal to 0\\nCairo traceback (most recent call last):\\nUnknown location (pc=0:3497)\\nUnknown location (pc=0:3480)\\nError message: Starkway: Token is no
t in
itialized\\nUnknown location (pc=0:3562)\\n\", source: CairoRunError(VmException(VmException { pc: 24, inst_location: None, inner_exc: Hint((0, AssertNotZero((0, \"0x8000000000000110000000
0000
0000000000000000000000000000000000001\")))), error_attr_value: None, traceback: Some(\"Cairo traceback (most recent call last):\\nUnknown location (pc=0:3497)\\nUnknown location (pc=0:3480
)\\n
Error message: Starkway: Token is not initialized\\nUnknown location (pc=0:3562)\\n\") })) })"}
github.com/NethermindEth/juno/junotest.getEstimate
        /home/user/repos/juno/junotest/estimate_message_fee_test.go:157
github.com/NethermindEth/juno/junotest.TestEstimateMessageFee.func2
        /home/user/repos/juno/junotest/estimate_message_fee_test.go:99
github.com/sourcegraph/conc/stream.(*Stream).Go.func1
        /home/user/.go/pkg/mod/github.com/sourcegraph/conc@v0.2.0/stream/stream.go:84
github.com/sourcegraph/conc/pool.(*Pool).worker
        /home/user/.go/pkg/mod/github.com/sourcegraph/conc@v0.2.0/pool/pool.go:137
github.com/sourcegraph/conc/panics.(*Catcher).Try
        /home/user/.go/pkg/mod/github.com/sourcegraph/conc@v0.2.0/panics/panics.go:23
github.com/sourcegraph/conc.(*WaitGroup).Go.func1
        /home/user/.go/pkg/mod/github.com/sourcegraph/conc@v0.2.0/waitgroup.go:32
18:29:58.444 19/07/2023 -07:00  ERROR   junotest/estimate_message_fee_test.go:157       Failed to estimate message fee  {"errCode": 40, "errMsg": "Contract error", "errData": "failed txn T
rans
actionHash(StarkFelt(\"0x03200d16a2d8ee1bd07edbf5c4c2b0c15193284d205e5d7cbceb022d780d861a\")) reason:ExecutionError(VirtualMachineExecutionErrorWithTrace { trace: \"Error in the called con
trac
t (0x05f5bc154f94a9d3cc892300eb872e5db0ad8cfa3ae8b5d202cea3b85407f7aa):\\nError at pc=0:24:\\nGot an exception while executing a hint: Assertion failed, 0 % 0x80000000000001100000000000000
0000
000000000000000000000000000001 is equal to 0\\nCairo traceback (most recent call last):\\nUnknown location (pc=0:3622)\\nUnknown location (pc=0:3605)\\nError message: Starkway: Token is no
t in
itialized\\nUnknown location (pc=0:3687)\\n\", source: CairoRunError(VmException(VmException { pc: 24, inst_location: None, inner_exc: Hint((0, AssertNotZero((0, \"0x8000000000000110000000
0000
0000000000000000000000000000000000001\")))), error_attr_value: None, traceback: Some(\"Cairo traceback (most recent call last):\\nUnknown location (pc=0:3622)\\nUnknown location (pc=0:3605
)\\n
Error message: Starkway: Token is not initialized\\nUnknown location (pc=0:3687)\\n\") })) })"}
github.com/NethermindEth/juno/junotest.getEstimate
        /home/user/repos/juno/junotest/estimate_message_fee_test.go:157
github.com/NethermindEth/juno/junotest.TestEstimateMessageFee.func2
        /home/user/repos/juno/junotest/estimate_message_fee_test.go:99
github.com/sourcegraph/conc/stream.(*Stream).Go.func1
        /home/user/.go/pkg/mod/github.com/sourcegraph/conc@v0.2.0/stream/stream.go:84
github.com/sourcegraph/conc/pool.(*Pool).worker
        /home/user/.go/pkg/mod/github.com/sourcegraph/conc@v0.2.0/pool/pool.go:137
github.com/sourcegraph/conc/panics.(*Catcher).Try
        /home/user/.go/pkg/mod/github.com/sourcegraph/conc@v0.2.0/panics/panics.go:23
github.com/sourcegraph/conc.(*WaitGroup).Go.func1
        /home/user/.go/pkg/mod/github.com/sourcegraph/conc@v0.2.0/waitgroup.go:32
18:32:08.470 19/07/2023 -07:00  ERROR   junotest/estimate_message_fee_test.go:157       Failed to estimate message fee  {"errCode": 40, "errMsg": "Contract error", "errData": "failed txn T
rans
actionHash(StarkFelt(\"0x03e42685d8cdca209dd0e4fd33c2c1922612152c441db96fa4cb1853674d30f1\")) reason:ExecutionError(VirtualMachineExecutionErrorWithTrace { trace: \"Error in the called con
trac
t (0x07e990a505f9fc7f0a8a279794e88d4f0e7abf6d8ed5b2891f6c56f5c38631e2):\\nError at pc=0:24:\\nGot an exception while executing a hint: Assertion failed, 0 % 0x80000000000001100000000000000
0000
000000000000000000000000000001 is equal to 0\\nCairo traceback (most recent call last):\\nUnknown location (pc=0:3606)\\nUnknown location (pc=0:3586)\\nError message: Starkway: Token is no
t in
itialized\\nUnknown location (pc=0:6147)\\n\", source: CairoRunError(VmException(VmException { pc: 24, inst_location: None, inner_exc: Hint((0, AssertNotZero((0, \"0x8000000000000110000000
0000
0000000000000000000000000000000000001\")))), error_attr_value: None, traceback: Some(\"Cairo traceback (most recent call last):\\nUnknown location (pc=0:3606)\\nUnknown location (pc=0:3586
)\\n
Error message: Starkway: Token is not initialized\\nUnknown location (pc=0:6147)\\n\") })) })"}
github.com/NethermindEth/juno/junotest.getEstimate
        /home/user/repos/juno/junotest/estimate_message_fee_test.go:157
github.com/NethermindEth/juno/junotest.TestEstimateMessageFee.func2
        /home/user/repos/juno/junotest/estimate_message_fee_test.go:99
github.com/sourcegraph/conc/stream.(*Stream).Go.func1
        /home/user/.go/pkg/mod/github.com/sourcegraph/conc@v0.2.0/stream/stream.go:84
github.com/sourcegraph/conc/pool.(*Pool).worker
        /home/user/.go/pkg/mod/github.com/sourcegraph/conc@v0.2.0/pool/pool.go:137
github.com/sourcegraph/conc/panics.(*Catcher).Try
        /home/user/.go/pkg/mod/github.com/sourcegraph/conc@v0.2.0/panics/panics.go:23
github.com/sourcegraph/conc.(*WaitGroup).Go.func1
        /home/user/.go/pkg/mod/github.com/sourcegraph/conc@v0.2.0/waitgroup.go:32
```

**Comment by joshklop** [comment]
I'm still not exactly sure how to tell if the estimates are correct, since no fee is actually charged for L1 handlers on L2 right now.

**Comment by joshklop** [comment]
Last testing TODO is to compare results with Pathfinder (and figure out what is causing those 8 errors, if we care to chase that down).

File: vm/vm.go [hunk 2]
```diff
@@ -117,7 +117,7 @@ func Call(contractAddr, selector *felt.Felt, calldata []felt.Felt, blockNumber,
 
 // Execute executes a given transaction set and returns the gas spent per transaction
 func Execute(txns []core.Transaction, declaredClasses []core.Class, blockNumber, blockTimestamp uint64,
-	sequencerAddress *felt.Felt, state core.StateReader, network utils.Network,
+	sequencerAddress *felt.Felt, state core.StateReader, network utils.Network, paidFeeOnL1 *felt.Felt,
```
Comment (omerfirmak) [inline vm/vm.go:120]: This is not possible for estimateMessageFee, but to keep `Execute` API suitable for all usecases of execution;

`txns` could contain multiple `L1Handlers`. Similar to `declaredClasses []core.Class` we need to have `paidFeesOnL1 []*felt.Felt`

**Comment by joshklop** [comment]
Confirmed that the last error in the above logs is for a transaction sent sometime before April, so all post-0.12 transactions seem to execute fine on Goerli2.

**Comment by joshklop** [comment]
@wojciechos found a bug for `estimateMessageFee` and `estimateFee` (panic due to nil pointer dereference when sequencer address is not present in the block). We now return errors in the JSONRPC response instead of panics in these cases.

Not worth it to do a ton of work to include sequencer address for anomalous (and very old) blocks. Would require a sizable migration.

File: vm/rust/src/lib.rs [hunk 3]
```diff
@@ -222,14 +216,13 @@ fn build_block_context(
...
     block_timestamp: c_ulonglong,
-    sequencer_address: StarkFelt,
 ) -> BlockContext {
     BlockContext {
         chain_id: ChainId(chain_id_str.into()),
         block_number: BlockNumber(block_number),
         block_timestamp: BlockTimestamp(block_timestamp),
 
-        sequencer_address: ContractAddress(PatriciaKey::try_from(sequencer_address).unwrap()),
+        sequencer_address: ContractAddress::default(),
```
Comment (omerfirmak) [inline vm/rust/src/lib.rs:313]: > Sequencer address is unnecessary for execution 

it actually is, fee contract asserts that `sequencer_address != 0x0`

We can use the fallback sequencer address, see https://github.com/NethermindEth/juno/pull/896#discussion_r1270767340

File: vm/rust/src/lib.rs [hunk 4]
```diff
@@ -177,15 +191,83 @@ pub extern "C" fn cairoVMExecute(
...
         }
 
         let res = match txn.unwrap() {
             Transaction::AccountTransaction(t) => t.execute(&mut state, &block_context),
-            Transaction::L1HandlerTransaction(t) => t.execute(&mut state, &block_context),
+            Transaction::L1HandlerTransaction(t) => {
+                // Manually inline L1HandlerTransaction.execute and execute_raw since the
+                // `actual_fee` in Blockifier is zero (i.e., Blockifier doesn't charge for L1
+                // Handler transactions on L2). This way we can get the full
+                // TransactionExecutionResult.
```
Comment (omerfirmak) [inline vm/rust/src/lib.rs:217]: Do we need to inline this entire thing?

`actual_resources` is a part of `TransactionExecutionInfo`. So we should be able to inline `calculate_tx_fee(&actual_resources, &context.block_context)?;` only?

File: vm/rust/src/lib.rs (hunk 3 above)
Comment (joshklop) [reply vm/rust/src/lib.rs:313]: Interesting. I wonder why it worked with my manual tests.

File: vm/rust/src/lib.rs (hunk 4 above)
Comment (joshklop) [reply vm/rust/src/lib.rs:216]: We get `actual_resources` from `calculate_tx_resources`, which requires `calculate_l1_gas_usage`, which requires `execute_call_info`, which requires `t.run_execute`, etc.

Is there a way to inline less? I haven't looked much into the implementation of the functions to see if they modify their inputs, of some of them take mutable references so I played it safe.

File: vm/rust/src/lib.rs [hunk 5]
```diff
@@ -226,14 +303,15 @@ fn build_block_context(
...
 ) -> BlockContext {
     BlockContext {
         chain_id: ChainId(chain_id_str.into()),
         block_number: BlockNumber(block_number),
         block_timestamp: BlockTimestamp(block_timestamp),
 
-        sequencer_address: ContractAddress(PatriciaKey::try_from(sequencer_address).unwrap()),
+        // Use default sequencer address
+        // See https://github.com/NethermindEth/juno/pull/901#discussion_r1271858429
+        sequencer_address: ContractAddress::try_from(StarkHash::try_from("0x046a89ae102987331d369645031b49c27738ed096f2789c24449966da4c6de6b").unwrap()).unwrap(),
```
Comment (omerfirmak) [inline vm/rust/src/lib.rs:314]: This should only be used as a fallback, not the default.
see https://github.com/NethermindEth/juno/pull/896/files#diff-1f724c5e433be46bfd8dc0962d1796f0dce2a5e22253fe82f69ab6a1eb8df278R1168-R1171

File: vm/rust/src/lib.rs (hunk 4 above)
Comment (omerfirmak) [reply vm/rust/src/lib.rs:216]: Correct me if I am wrong but I believe this works too

``` rust
        let res = match txn.unwrap() {
            Transaction::AccountTransaction(t) => t.execute(&mut state, &block_context),
            Transaction::L1HandlerTransaction(t) => {
                let maybe_execution_info = t.execute(&mut state, &block_context);
                if maybe_execution_info.is_err() {
                    maybe_execution_info
                } else {
                    let mut execution_info = maybe_execution_info.unwrap();
                    execution_info.actual_fee = calculate_tx_fee(&execution_info.actual_resources, &block_context).unwrap();
                    Ok(execution_info)    
                }
            },
        };
```

File: vm/vm.go [hunk 6]
```diff
@@ -133,6 +134,13 @@ func Execute(txns []core.Transaction, declaredClasses []core.Class, blockNumber,
 	txnsJSONCstr := C.CString(string(txnsJSON))
 	classesJSONCStr := C.CString(string(classesJSON))
 
+	if sequencerAddress == nil {
+		// Set to fallback address
+		sequencerAddress, err = new(felt.Felt).SetString("0x046a89ae102987331d369645031b49c27738ed096f2789c24449966da4c6de6b")
+		if err != nil {
+			return nil, fmt.Errorf("set sequencer address to fallback address: %w", err)
+		}
+	}
```
Comment (omerfirmak) [inline vm/vm.go:149]: :)

https://github.com/NethermindEth/juno/pull/896#discussion_r1270767340

Comment (joshklop) [reply vm/vm.go:149]: XD

Will fix, thanks.

**Comment by omerfirmak** [comment]
Overall this looks good to me. I will try to get it across the finish line since I believe @joshklop is less available today.

**Review by omerfirmak — APPROVED** [review]
I will wait for someone else to approve as well and @joshklop to sign off on the changes

File: rpc/handlers_test.go [hunk 7]
```diff
@@ -2112,3 +2113,64 @@ func TestCall(t *testing.T) {
...
+		Timestamp: 456,
+		GasPrice:  new(felt.Felt).SetUint64(42),
+	}
+	mockState := mocks.NewMockStateHistoryReader(mockCtrl)
+	NoopCloser := func() error { return nil }
+
+	mockReader.EXPECT().HeadState().Return(mockState, NoopCloser, nil)
+	mockReader.EXPECT().HeadsHeader().Return(latestHeader, nil)
+
+	expectedGasConsumed := new(felt.Felt).SetUint64(37)
```
Comment (joshklop) [inline rpc/handlers_test.go:2150]: How did you calculate this?

Comment (joshklop) [reply rpc/handlers_test.go:2150]: Same question for the other values. I hadn't figured out how to test if the values themselves are correct, so if you did that would be amazing!

Comment (omerfirmak) [reply rpc/handlers_test.go:2150]: Nope that is a random value. We are using a mock vm in that test, so no real execution is going on.

**Comment by joshklop** [comment]
PR looks good to me. Thanks for making the changes @omerfirmak. I'm not in a position to develop right now, so I can't resolve the last comment (about the json). I can do it tomorrow. If you'd rather merge before then, feel free.

**Comment by omerfirmak** [comment]
> PR looks good to me. Thanks for making the changes @omerfirmak. I'm not in a position to develop right now, so I can't resolve the last comment (about the json). I can do it tomorrow. If you'd rather merge before then, feel free.

Nah, it is fine. It wasnt really an issue anyways. Will merge soon.

**Pull request merged**